docs/runtime/dotnet-metrics.md

```diff
@@ -0,0 +1,788 @@
+<!--- Hugo front matter used to generate the website version of this page:
+linkTitle: .NET
+--->
+
+# Semantic Conventions for .NET (CLR) Metrics
+
+**Status**: [Experimental][DocumentStatus]
+
+This document describes semantic conventions for .NET CLR runtime metrics in OpenTelemetry.
+
+<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->
+
+<!-- toc -->
+
+- [CLR Garbage Collection](#clr-garbage-collection)
+  - [Metric: `dotnet.gc.collections.count`](#metric-dotnetgccollectionscount)
+  - [Metric: `dotnet.gc.objects.size`](#metric-dotnetgcobjectssize)
+  - [Metric: `dotnet.gc.memory.total_allocated`](#metric-dotnetgcmemorytotal_allocated)
+  - [Metric: `dotnet.gc.memory.commited`](#metric-dotnetgcmemorycommited)
```

Comment:
@lmolkova - were you interested in pursuing dotnet.memory.* or dotnet.heap.* ? That seemed like our biggest open design question right now.

Thanks for all the feedback!

Reply:
Not really interested in `dotnet.memory.*`  - it's too abstract. 

I still don't understand where we'd report non-gc heap - https://github.com/dotnet/diagnostics/issues/4156

I want to understand how `dotnet.heap` can evolve and ideally group metrics that could be grouped under the same name.

It could either be 
- `dotnet.heap.size` with an extra dimension for `heap_type = gc|non-gc`
- another metric `dotnet.non_gc.heap.*`
- or it'd be reported as a part of something else.

Reply:
Just to make sure we are on the same page, in .NET 9 we have no easy API to get the size of the non-gc heap so I assume all discussion about it is how would we adjust the metrics in some future release to add it. The non-gc heap also tends to be pretty small so its probably not a high value metric on its own, but its a fine example of the general principle that .NET allocates other heaps besides just the GC. It might be useful to show users the size of some of them in the future. Jitted code is an example of another heap that could be bigger and potentially more useful.

If we do dotnet.heap.size with a dimension for heap_type that works, but it gives a weird meaning for the overall metric when not filtered by any attributes. For example assume we shipped .NET 9 where heap_type=gc is the only value, then in .NET 10 we added non-gc and jitted_code to the possible values. Everyone who graphs dotnet.heap.size will see the value jump up as soon as they upgrade to .NET 10 because now we are reporting on more heaps than we did before. I suspect lots of customers would misinterpret and think "oh no, .NET 10 uses so much more memory" when it was just reporting that changed.

We could solve that problem by adding heap_type=Other and the value of Other would be all other committed memory in process that is not part of a specially reported heap. This would mean dotnet.heap.size = total_committed_process memory, and we may as well rename it dotnet.memory. That actually didn't seem problematic to me, but happy to discuss because you didn't seem to favor it.

The other way I could see doing this is that we don't have any metric that tries to aggregate heaps together and instead each heap gets its own metric. So over time we might have:
dotnet.heap.gc.size
dotnet.heap.non-gc.size
dotnet.heap.jitted_code.size
dotnet.heap.loader.size
...

I'm OK with that too, but it does have the problem you alluded to earlier that if you are trying to start from high level memory usage and break it down then you need to be aware of multiple different metrics to compose the story together.

Reply:
Thanks for the context @noahfalk 

Yes, my concern was for the future releases. 

To summarize options:
1. `dotnet.memory.*` or `dotnet.heap`
   - It would be easier to start with (providing high-level break-down for memory types and sub-types). 
   - The downside is that we could end up with multiple sub-type attributes that only make sense for a specific type (gc|jitted_code|...). It's hard to explain what this metric means without filters 
   - we'd still need other GC-memory metrics like `dotnet.gc.heap.fragmentation` or `dotnet.gc.objects.size`. 
2.  Keep independent `dotnet.heap.gc`, `dotnet.heap.non-gc`,  `dotnet.heap.jitted_code`, .. metrics
   - it's precise and we can easily evolve metrics without changing anything for existing ones or making them too abstract
   - it maps better to existing .NET GC APIs and event (perf) counters
   - it has a bit steeper learning curve to know what to look for

My preference would be on Option 2 + better docs
- we can provide dashboards + docs on what to monitor (assuming users don't have deep knowledge of .NET memory management). 
- we can explain how `dotnet.gc.memory.commited` is related to heap size and that there are  other metrics we don't capture yet

If we can settle on it, I'd like to circle back on naming and discuss
- `dotnet.heap.gc.*` instead of `dotnet.gc.heap.*`
- do we need `dotnet.gc.memory` or corresponding metrics should also be in `dotnet.heap.gc.` namespace

Reply:
I'm happy with either option (1) or (2) above, and since you prefer (2) lets take that as our plan of record. Given its more aligned with what the previous OTel metrics were doing I'm not expecting novel issues to arise.

> I'd like to circle back on naming and discuss

Sure. For the dotnet.heap.gc.\* vs. dotnet.gc.heap.\*, I'm guessing your goal was to bring together different heaps so that they would all start with dotnet.heap.* like in my example? I'm realizing in hindsight that may be an awkward pattern. This is how some metrics would look like under three different patterns:

**1. Insert heap before feature name for all metrics in a feature that uses a heap**
`dotnet.heap.gc.size`
`dotnet.heap.gc.fragmentation`
`dotnet.heap.gc.time`
`dotnet.heap.jit.size` (hypothetical)
`dotnet.heap.jit.compiled_il.size`
`dotnet.heap.jit.compiled_methods`
`dotnet.heap.jit.compilation.time`
`dotnet.heap.assembly.count`
`dotnet.heap.assembly.size` (hypothetical)

**2. Insert heap before feature name only for heap related metrics**
`dotnet.heap.gc.size`
`dotnet.heap.gc.fragmentation`
`dotnet.gc.time`
`dotnet.heap.jit.size` (hypothetical)
`dotnet.jit.compiled_il.size`
`dotnet.jit.compiled_methods`
`dotnet.jit.compilation.time`
`dotnet.assembly.count`
`dotnet.heap.assembly.size` (hypothetical)

**3. Insert heap after feature name for heap related metrics**
`dotnet.gc.heap.size`
`dotnet.gc.heap.fragmentation`
`dotnet.gc.time`
`dotnet.jit.heap.size` (hypothetical)
`dotnet.jit.compiled_il.size`
`dotnet.jit.compiled_methods`
`dotnet.jit.compilation.time`
`dotnet.assembly.count`
`dotnet.assembly.heap.size` (hypothetical)

IMO (3) feels the most natural and organized, but it doesn't have the property that you can enumerate every metric that starts with dotnet.heap to understand what all the heaps are. Doing (1) is hard to do correctly because you have to anticipate every feature where you might want to add a heap size metric in the future and add the 'heap' name in up-front. (2) is more manageable, but then we've got both dotnet.heap.feature and dotnet.feature every time we add a heap metric.


> do we need dotnet.gc.memory or corresponding metrics should also be in dotnet.heap.gc. namespace

I agree that having both 'memory' and 'heap' seems a bit arbitrary. If we went with option (3) above perhaps:

`dotnet.gc.memory.total_allocated` becomes `dotnet.gc.heap.total_allocated`
`dotnet.gc.memory.commited` would go away as a metric name and the value would be reported by 'dotnet.heap.size' metric. This would require we add an 'Other' or 'None' value to the generation attribute and use that to report the committed memory that isn't associated with a specific generation.

Reply:
Option 3 sounds good to me, thanks a lot for summarizing the proposal.

Reply:
> I agree that having both 'memory' and 'heap' seems a bit arbitrary. If we went with option (3) above perhaps ...

After discussing with Maoni it turns out 'heap' and 'memory' aren't as arbitrary as I first believed. Maoni says that in the domain for .NET GC the term 'heap' has a precise meaning. ".NET GC heap" includes all the memory used for storing .NET objects and the small fragmentation gaps in between, but it does not include the additional pool of committed memory which is kept cached to grow the heap on demand. So if we apply the terminology precisely dotnet.gc.heap.total_allocated is fine. However refering to GC committed memory as dotnet.gc.heap.committed or dotnet.heap.size would both be confusing options. The terms imply that they are only refering to the 'heap', but the values reported would include committed memory from the cache which is not considered part of the heap.

Reply:
> dotnet.memory.*

I want to revisit this again not from the perspective of making a hierarchial break-down of .NET memory usage, but just from trying to achieve parity with the existing EventCounter "working-set". Today in .NET the "working-set" counter reports the value of the Environment.WorkingSet API. I think of it as a rough analog of dotnet.cpu.time except for memory instead of CPU usage. Does a dotnet.memory.working_set metric with no dimensions seem objectionable?

Given that OTel's existing .NET metrics don't have it we might be able to get by without it, but I expect people will ask for it.